projects/packages/connection/src/class-external-storage.php

```diff
@@ -0,0 +1,237 @@
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param string $option_name The option name to retrieve.
+	 * @return mixed The option value from external storage, or null for database fallback.
+	 */
+	public static function get_option( $option_name ) {
+		$provider = self::$provider;
+
+		// Check if we have a registered provider
+		if ( null === $provider ) {
+			return null; // No provider registered, use database
+		}
+
+		// Get environment ID from provider
+		$environment = method_exists( $provider, 'get_environment_id' ) ? $provider->get_environment_id() : 'unknown';
+
+		// Check if provider is available in current environment
+		if ( ! $provider->is_available() ) {
+			self::log_event( 'unavailable', $option_name, 'External storage not available', $environment );
+			return null;
+		}
+
+		// Check if provider should handle this option
+		if ( ! $provider->should_handle( $option_name ) ) {
+			return null;
+		}
+
+		// Try to get value from the provider
+		try {
+			$value = $provider->get( $option_name );
+
+			// Check if we got a valid value (excluding null, false, empty string, and zero)
+			if ( null !== $value && false !== $value && '' !== $value && 0 !== $value ) {
+				return $value;
+			}
+
+			// Empty value - log it
+			self::log_event( 'empty', $option_name, '', $environment );
+
+		} catch ( \Exception $e ) {
+			// Provider threw an exception
+			self::log_event( 'error', $option_name, $e->getMessage(), $environment );
+		}
+
+		// Provider couldn't provide value, return null for database fallback
+		return null;
+	}
+
+	/**
+	 * Log external storage events to the Jetpack Connection Error_Handler.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param string $event_type  The event type (error, empty, unavailable).
+	 * @param string $option_name The option name that triggered the event.
+	 * @param string $details     Additional details about the event.
+	 * @param string $environment The environment identifier (atomic, vip, etc.).
+	 */
+	public static function log_event( $event_type, $option_name, $details = '', $environment = 'unknown' ) {
+		$should_log = false;
+
+		if ( 'error' === $event_type ) {
+			// Report external storage errors for supported environments
+			$should_log = self::should_report_for_environment();
+		} elseif ( 'empty' === $event_type ) {
+			// Use delay mechanism to distinguish disconnection from a delay
+			$should_log = self::should_report_for_environment() && self::should_report_empty_state( $option_name );
+		} elseif ( 'unavailable' === $event_type ) {
+			// Log locally but don't report to WordPress.com
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+				error_log( sprintf( 'External storage unavailable: %s in %s%s', $option_name, $environment, $details ? ' - ' . $details : '' ) ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+			}
+			return;
+		}
+
+		if ( ! $should_log || ! class_exists( 'Automattic\Jetpack\Connection\Error_Handler' ) ) {
+			return;
+		}
```

Comment:
It's probably a matter of personal taste, but I feel like the code could be simplified a bit.

```suggestion
		if ( 'unavailable' === $event_type ) {
			// Log locally but don't report to WordPress.com
			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
				error_log( sprintf( 'External storage unavailable: %s in %s%s', $option_name, $environment, $details ? ' - ' . $details : '' ) ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
			}
			return;
		}

		if ( ! self::should_report_for_environment() ) {
			return;
		}

		if ( 'empty' === $event_type && ! self::should_report_empty_state( $option_name ) ) {
			// Use delay mechanism to distinguish disconnection from a delay
			return;
		}

		if (  ! class_exists( 'Automattic\Jetpack\Connection\Error_Handler' ) ) {
			return;
		}
```

Reply:
Took at fresh look at this and yeah, that's entirely personal taste. I just like to return early if I can.
Please disregard.